libsrc/_DEVELOPMENT/l/sccz80/4-ez80/lp_gchar.asm

```diff
@@ -4,18 +4,16 @@
 ; Entry: EHL=far pointer
 ; Exit: a=byte
 
+SECTION code_clib
+SECTION code_l_sccz80
+
+.ASSUME ADL = 0        ; use the 16 bit address, with MBASE
+
 lp_gchar:
-    defb    $5b    ;lil
-    push    hl
-    defb    $5b    ;lil
-    ld      hl,2
-    defb    0
-    defb    $5b
-    add     hl,sp
-    defb    $5b
-    ld      (hl),e
-    defb    $5b
-    pop     hl
-    defb    $5b
-    ld      a,(hl)
-    ret
+    push.l    hl       ; push 3 bytes (HLU is undefined in Z80 MEMORY Mode)
+    ld.lil    hl,2     ; point to pushed HLU by loading 16 bit 0x0002 to HL, HLU is written to 0x00.
+    add.l     hl,sp    ; by adding the 24 bit SPL to 24 bit HL
+    ld.l      (hl),e   ; store the E register at the HLU location in stack
+    pop.l     hl       ; get a correct (far pointer) long address into 24 bit HL
+    ld.l      a,(hl)   ; get a byte from that location into A.
+    ret.l              ; return correctly from a long call
```

Comment:
These functions are called from z80 side, so this should be a short ret?

Reply:
You're right if it can be guaranteed that it will only be called from a z80 short call.

But the guidelines in the datasheet suggest...

4. Any code block that may be called from either Z80 mode or ADL mode must be called with a suffix to save the calling code’s memory mode on the SPL stack.
5. Any routine that may be called from either mode must return with a suffixed `RET.L` instruction to restore the calling code’s memory mode from the SPL stack.

I don't know how these will be used, but if they're intrinsic couldn't the be called from code running in either mode?  :man_shrugging:
But doing so would break everything else, unless everything was assembled in ADL mode.

I'll change it when we find out about the `ld.lis rr,mm` correctness (or not).